distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/NodesRoutingTable.java

```diff
@@ -0,0 +1,154 @@
+ * Size of RoutingTable is fixed and equals to nodeOwner's key length
+ * @author Giorgia Bortoletti
+ */
+public class NodesRoutingTable extends RoutingTable<KBucket> {
+
+    private List<KBucket> bucketsTable;
+    private int sizeTable;
+    private Node nodeOwner;
+
+    /**
+     * Constructor where the routing table lenght is equal to peer owner length
+     * @param nodeOwner
+     */
+    public NodesRoutingTable(Node nodeOwner){
+        new NodesRoutingTable(nodeOwner, nodeOwner.keyLength());
+    }
+
+    /**
+     * Constructor where the routing table lenght is sizeTable
+     * @param nodeOwner
+     * @param sizeTable dimension of routing table
+     */
+    public NodesRoutingTable(Node nodeOwner, int sizeTable){
+        this.nodeOwner = nodeOwner;
+        this.sizeTable = sizeTable;
+        bucketsTable = new ArrayList<>(sizeTable);
+        for(int i=0; i<sizeTable; i++)
+            bucketsTable.add(new KBucket(sizeTable));
+    }
+
+    /**
+     * @return nodeOwner of this routing table
+     */
+    public Node getNodeOwner(){ return nodeOwner; }
+
+    /**
+     * @param node Node to add
+     * @return true if the node has been added, false otherwise
+     */
+    @Override
+    public boolean add(Node node) {
+        if(node == null)
+            return false;
+        return bucketsTable.get(getLocation(node)).add(node);
+    }
+
+    /**
+     * @param node Node to remove
+     * @return true if the node has been removed, false otherwise
+     */
+    @Override
+    public boolean remove(Node node) {
+        if(node == null)
+            return false;
+        int positionNode = getLocation(node);
+        KBucket bucketFound = bucketsTable.get(positionNode);
+        if(!bucketFound.contains(node))
+            return false;
+        return bucketFound.remove(node);
+    }
+
+    /**
+     * @param node Node of which check presence
+     * @return true if present, false otherwise
+     */
+    @Override
+    public boolean contains(Node node) {
+        if(node == null || bucketsTable.size() == 0)
+            return false;
+        int positionNode = getLocation(node);
+        if(positionNode == -1)
+            return false;
+        return bucketsTable.get(positionNode).contains(node);
+    }
+
+    /**
+     * @param i index of the bucket in buckets container
+     * @return the bucket at index i
+     */
+    @Override
+    public KBucket getBucket(int i) {
+        try {
+            return bucketsTable.get(i);
+        }catch (IndexOutOfBoundsException e){
+            return null;
+        }
+    }
+
+    /**
+     * @param node Node of the distributed Network
+     * @return the index (between 0 and N -1) of the bucket that maybe containing the given Node with BinarySet, -1 otherwise
+     */
+    @Override
+    public int getLocation(Node node) {
+        if(node == null)
+            return -1;
+        Object distanceObject = nodeOwner.getDistance(node);
```

Comment:
nodeOwner is already a Node but with a generic type so I must have the certainty of this generic type is BinarySet. Otherwise I should declare nodeOwner as Node<BinaryKey> and only in the KBucket.
Do you think this replacement could be the best solution or do I always check that if the distance is a Binary Set?

Reply:
Well, it does not make a lot of sense to work with generic Nodes in the routing table since it's based on binary addresses. Maybe it should accept Node<BinarySet> only?

Reply:
I think it's fair that the routing table only accepts Node<BinarySet>, now I fix it.